src/pages/permissions/panels/EditIdentityGroupsPanel.tsx

```diff
@@ -0,0 +1,350 @@
+import {
+  ActionButton,
+  Button,
+  Icon,
+  useNotify,
+} from "@canonical/react-components";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { fetchLxdGroups } from "api/auth-groups";
+import SidePanel from "components/SidePanel";
+import { FC, useEffect, useState } from "react";
+import { queryKeys } from "util/queryKeys";
+import usePanelParams from "util/usePanelParams";
+import PermissionLxdGroupsFilter, {
+  PermissionLxdGroupsFilterType,
+  QUERY,
+} from "../PermissionLxdGroupsFilter";
+import ScrollableTable from "components/ScrollableTable";
+import SelectableMainTable from "components/SelectableMainTable";
+import { useSearchParams } from "react-router-dom";
+import {
+  generateGroupAllocationsForIdentities,
+  getCurrentGroupAllocationsForIdentities,
+} from "util/permissionIdentities";
+import useSortTableData from "util/useSortTableData";
+import { useToastNotification } from "context/toastNotificationProvider";
+import useEditHistory from "util/useEditHistory";
+import IdentityGroupsPanelActions from "./IdentityGroupsPanelActions";
+import { pluralize } from "util/instanceBulkActions";
+import IdentityGroupsPanelConfirmModal from "./IdentityGroupsPanelConfirmModal";
+import { updateGroupsForIdentities } from "api/auth-identities";
+import { LxdIdentity } from "types/permissions";
+
+type PanelHistoryState = {
+  groupsAssignedToEveryIdentity: string[];
+  groupsAssignedToSomeIdentities: string[];
+  groupsAssignedToNoIdentities: string[];
+};
+
+interface Props {
+  selectedIdentities: LxdIdentity[];
+}
+
+const EditIdentityGroupsPanel: FC<Props> = ({ selectedIdentities }) => {
+  const panelParams = usePanelParams();
+  const [searchParams] = useSearchParams();
+  const notify = useNotify();
+  const toastNotify = useToastNotification();
+  const queryClient = useQueryClient();
+  const [submitting, setSubmitting] = useState(false);
+  const [confirming, setConfirming] = useState(false);
+
+  const {
+    data: groups = [],
+    error,
+    isLoading,
+  } = useQuery({
+    queryKey: [queryKeys.lxdGroups],
+    queryFn: fetchLxdGroups,
+  });
+
+  const {
+    desiredState,
+    save: saveToPanelHistory,
+    undo: undoGroupChange,
+  } = useEditHistory<PanelHistoryState>({
+    initialState: {
+      groupsAssignedToEveryIdentity: [],
+      groupsAssignedToSomeIdentities: [],
+      groupsAssignedToNoIdentities: [],
+    },
+  });
+
+  if (error) {
+    notify.failure("Loading panel details failed", error);
+  }
+
+  // Figure out if groups are allocated to all or some of the selected identities when the panel initially mounts
+  useEffect(() => {
```

Comment:
This is needed because we keep all the table data states in the undo stack (`useEditHistory`). We can't initialise the undo stack with the group allocations before the first render cycle because at that point we still need to query for the groups data.

Reply:
I think this is because the way state is kept is a bit too complex. When we only keep the user interactions (removed/added groups) in state, we can also only keep those in the undo stack and the stack initializes empty.

Reply:
Yeah you are right, this is likely to change once I have refactored the code in this component a bit :+1:

Reply:
This is now gone.